// @azure-rest/core-client/dist/commonjs/restError.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRestError = void 0;
const core_rest_pipeline_1 = require("@azure/core-rest-pipeline");
function createRestError(messageOrResponse, response) {
    const resp = typeof messageOrResponse === "string" ? response : messageOrResponse;
    const internalError = resp.body.error || resp.body;
    const message = typeof messageOrResponse === "string"
        ? messageOrResponse
        : internalError.message ?? `Unexpected status code: ${resp.status}`;
    return new core_rest_pipeline_1.RestError(message, {
        statusCode: statusCodeToNumber(resp.status),
        code: internalError.code,
        request: resp.request,
        response: toPipelineResponse(resp),
    });
}
exports.createRestError = createRestError;
function toPipelineResponse(response) {
    return {
        headers: (0, core_rest_pipeline_1.createHttpHeaders)(response.headers),
        request: response.request,
        status: statusCodeToNumber(response.status) ?? -1,
    };
}
function statusCodeToNumber(statusCode) {
    const status = Number.parseInt(statusCode);
    return Number.isNaN(status) ? undefined : status;
}
