// @azure/openai/dist-esm/src/classic/client/openAIClient/index.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { getAudioTranscriptionAsPlainText, getAudioTranscriptionAsResponseObject, getAudioTranslationAsPlainText, getAudioTranslationAsResponseObject, getCompletions, getChatCompletions, getChatCompletionsWithAzureExtensions, getImageGenerations, getEmbeddings, } from "../../../api/client/openAIClient/index.js";
export function getClientOpenAIClient(context) {
    return {
        getAudioTranscriptionAsPlainText: (deploymentId, body, options) => getAudioTranscriptionAsPlainText(context, deploymentId, body, options),
        getAudioTranscriptionAsResponseObject: (deploymentId, body, options) => getAudioTranscriptionAsResponseObject(context, deploymentId, body, options),
        getAudioTranslationAsPlainText: (deploymentId, body, options) => getAudioTranslationAsPlainText(context, deploymentId, body, options),
        getAudioTranslationAsResponseObject: (deploymentId, body, options) => getAudioTranslationAsResponseObject(context, deploymentId, body, options),
        getCompletions: (deploymentId, body, options) => getCompletions(context, deploymentId, body.prompt, Object.assign(Object.assign({}, body), options)),
        getChatCompletions: (deploymentId, body, options) => getChatCompletions(context, deploymentId, body.messages, Object.assign(Object.assign({}, body), options)),
        getChatCompletionsWithAzureExtensions: (deploymentId, body, options) => getChatCompletionsWithAzureExtensions(context, deploymentId, body, options),
        getImageGenerations: (deploymentId, body, options) => getImageGenerations(context, deploymentId, body, options),
        getEmbeddings: (deploymentId, body, options) => getEmbeddings(context, deploymentId, body, options),
    };
}
export function getClientOpenAIClientOperations(context) {
    return Object.assign({}, getClientOpenAIClient(context));
}
